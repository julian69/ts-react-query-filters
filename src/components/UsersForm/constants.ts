import { FILTERS } from "@/constants/filters";
import { IFilters } from "@/types/users";

export const filtersInitialState: IFilters = {
  [FILTERS.NAME]: '',
  [FILTERS.EMAIL]: '',
  [FILTERS.COMPANY]: '',
  [FILTERS.ZIPCODE]: '',
};