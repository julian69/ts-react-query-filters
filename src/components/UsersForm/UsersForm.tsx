import { FC } from 'react';

import useUsersForm from './hooks/useUsersForm';
import { FILTERS } from '@/constants/filters';
import Button from '@/components/Button/Button';
import InputText from '@/components/InputText/InputText';

const UsersForm: FC = () => {
  const { filters, handleSubmit, handleReset, handleFiltersChange } = useUsersForm();

  return (
    <form className="rounded-xl md:border border-gray-200 bg-white p-6 pt-0 mb-10 mt-3 md:shadow-lg">
      <header className="py-4 flex justify-center md:justify-start">
        <h2 className="font-bold">Enter user details</h2>
      </header>
      <div className="grid grid-cols-1 gap-4 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4">
        <InputText
          label={'Name'}
          value={filters[FILTERS.NAME]}
          onChange={value => handleFiltersChange(FILTERS.NAME, value)} />
        <InputText
          label={'Email'}
          value={filters[FILTERS.EMAIL]}
          onChange={value => handleFiltersChange(FILTERS.EMAIL, value)} />
        <InputText
          label={'Company'}
          value={filters[FILTERS.COMPANY]}
          onChange={value => handleFiltersChange(FILTERS.COMPANY, value)} />
        <InputText
          label={'Zip Code'}
          value={filters[FILTERS.ZIPCODE]}
          onChange={value => handleFiltersChange(FILTERS.ZIPCODE, value)} />
      </div>
      <footer className="mt-6 grid w-full grid-cols-2 justify-end space-x-4 md:flex pt-4">
        <Button
          value="Reset"
          onClick={handleReset}
          color="bg-gray-200 text-gray-700"
          isDisabled={!Object.values(filters).some(Boolean)} />
        <Button
          value="Find Users"
          onClick={handleSubmit}
          isDisabled={!Object.values(filters).some(Boolean)} />
      </footer>
    </form>
  );
};

export default UsersForm;
