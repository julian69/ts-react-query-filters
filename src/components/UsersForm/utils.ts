import { IFilters } from "@/types/users";
import { searchParams } from "./constants";

export const deserializeUrl = (params: URLSearchParams): IFilters => ({
  name: params.get('name') || '',
  email: params.get('email') || '',
  company: params.get('company') || '',
  zipcode: params.get('zipcode') || '',
});

export const setSearchParams = (filters?: IFilters) => {
  if (!filters) {
    window.history.pushState(null, '', '/');
    return;
  }

  Object.keys(filters).forEach((key) => {
    searchParams.has(key)
      ? searchParams.set(key, filters[key])
      : searchParams.append(key, filters[key]);
  });

  window.history.pushState(null, '', `?${searchParams.toString()}`);
};