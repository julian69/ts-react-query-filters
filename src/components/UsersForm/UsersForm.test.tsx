import '@testing-library/jest-dom';
import { describe, expect, it } from 'vitest';

import { fireEvent, render, screen } from '@/test/utils';
import UsersForm from "@/components/UsersForm/UsersForm";
import { deserializeUrl, setSearchParams } from './utils';

describe('<UsersForm />', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<UsersForm />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should enable buttons on change', () => {
    render(<UsersForm />);
    const inputName = screen.getByLabelText('Name');
    const submit = screen.getByText('Find Users');
    const reset = screen.getByText('Find Users');

    expect(submit).toBeDisabled();
    expect(reset).toBeDisabled();

    fireEvent.change(inputName, { target: { value: '23' } });

    expect(submit).not.toBeDisabled();
    expect(reset).not.toBeDisabled();
  });

  it('should reset form when "Reset" button is clicked', () => {
    render(<UsersForm />);
    const input = screen.getByLabelText('Name');

    fireEvent.change(input, { target: { value: 'gra' } });

    const resetButton = screen.getByText('Reset');
    const nameInput = screen.getByLabelText('Name') as HTMLInputElement;

    fireEvent.click(resetButton);
    expect(nameInput.value).toBe('');
  });
});

describe('deserializeUrl', () => {
  it('should deserialize URLSearchParams into IFilters object', () => {
    const mockParams = new URLSearchParams("name=John&email=john@example.com&company=ABC&zipcode=12345");
    const result = deserializeUrl(mockParams);

    expect(result).toEqual({
      name: 'John',
      email: 'john@example.com',
      company: 'ABC',
      zipcode: '12345',
    });
  });

  it('should handle missing parameters', () => {
    const mockParams = new URLSearchParams("name=John&zipcode=12345");
    const result = deserializeUrl(mockParams);

    expect(result).toEqual({
      name: 'John',
      email: '',
      company: '',
      zipcode: '12345',
    });
  });
});

describe('setSearchParams', () => {
  it('should set search params in window history', () => {
    const mockFilters = {
      name: 'John',
      email: 'john@example.com',
      company: 'ABC',
      zipcode: '12345',
    };
    setSearchParams(mockFilters);
    expect(window.location.search).toBe('?name=John&email=john%40example.com&company=ABC&zipcode=12345');
  });

  it('should handle undefined filters', () => {
    setSearchParams();
    expect(window.location.search).toBe('');
  });
});