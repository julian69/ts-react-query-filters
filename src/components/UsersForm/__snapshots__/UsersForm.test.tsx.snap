// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<UsersForm /> > should render correctly 1`] = `
<DocumentFragment>
  <form
    class="rounded-xl md:border border-gray-200 bg-white p-6 pt-0 mb-10 mt-3 md:shadow-lg"
  >
    <header
      class="py-4 flex justify-center md:justify-start"
    >
      <h2
        class="font-bold"
      >
        Enter user details
      </h2>
    </header>
    <div
      class="grid grid-cols-1 gap-4 md:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4"
    >
      <div
        class="mr-5"
      >
        <label
          class="text-sm font-medium text-stone-600"
          for="name"
        >
          Name
        </label>
        <input
          class="mt-2 block w-full rounded-md border border-gray-100 bg-gray-100 px-2 py-2 shadow-sm outline-none focus:border-blue-500 focus:ring focus:ring-blue-200 focus:ring-opacity-50"
          id="name"
          type="text"
          value=""
        />
      </div>
      <div
        class="mr-5"
      >
        <label
          class="text-sm font-medium text-stone-600"
          for="email"
        >
          Email
        </label>
        <input
          class="mt-2 block w-full rounded-md border border-gray-100 bg-gray-100 px-2 py-2 shadow-sm outline-none focus:border-blue-500 focus:ring focus:ring-blue-200 focus:ring-opacity-50"
          id="email"
          type="text"
          value=""
        />
      </div>
      <div
        class="mr-5"
      >
        <label
          class="text-sm font-medium text-stone-600"
          for="company"
        >
          Company
        </label>
        <input
          class="mt-2 block w-full rounded-md border border-gray-100 bg-gray-100 px-2 py-2 shadow-sm outline-none focus:border-blue-500 focus:ring focus:ring-blue-200 focus:ring-opacity-50"
          id="company"
          type="text"
          value=""
        />
      </div>
      <div
        class="mr-5"
      >
        <label
          class="text-sm font-medium text-stone-600"
          for="zip,code"
        >
          Zip Code
        </label>
        <input
          class="mt-2 block w-full rounded-md border border-gray-100 bg-gray-100 px-2 py-2 shadow-sm outline-none focus:border-blue-500 focus:ring focus:ring-blue-200 focus:ring-opacity-50"
          id="zip,code"
          type="text"
          value=""
        />
      </div>
    </div>
    <footer
      class="mt-6 grid w-full grid-cols-2 justify-end space-x-4 md:flex pt-4"
    >
      <button
        class="rounded-lg bg-gray-200 text-gray-700 text-sm px-8 py-2 font-medium  outline-none hover:opacity-80 focus:ring" opacity-50"
        disabled=""
        type="submit"
      >
        Reset
      </button>
      <button
        class="rounded-lg bg-blue-600 text-white text-sm px-8 py-2 font-medium  outline-none hover:opacity-80 focus:ring" opacity-50"
        disabled=""
        type="submit"
      >
        Find Users
      </button>
    </footer>
  </form>
</DocumentFragment>
`;
